gk/main.c

```diff
@@ -584,15 +709,40 @@ alloc_gk_conf(void)
 }
 
 int
+gk_conf_hold(struct gk_config *gk_conf)
+{
+	if (gk_conf) {
```

Comment:
I see.

We don't want to check the parameters everywhere because it slows code down, and soon there will be checks after checks. It's essential to sanitize input data, but once we reach a state in which all data is sanitized, we avoid checking things. We also check calls coming from Lua because we assume that the Lua code is less robust than the core of Gatekeeper. Finally, asserts constitute another class of checks since they are meant to verify invariants of the code.

What I wrote above doesn't apply for useful exceptions. The most common useful exception that I can remember is that `free()` accepts `NULL` as a no operation. The reason for this exception is because it's really useful. For example, consider a pointer to a string. The `NULL` can be used a compact representation of the empty string.